src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy

```diff
@@ -51,32 +54,121 @@ class TeamCityBuildScanPluginTest extends Specification {
-        buildScript.text = """
+    def "service messages emitted when build scan plugin is applied later than the plugin"() {
+        given:
+        configureBuildScanPlugin(BUILD_SCAN_PLUGIN_CLASSPATH)
+        buildScript << """
             plugins {
-                id 'com.gradle.build-scan' version '${version}'
                 id 'nu.studer.build-scan.teamcity'
+                id 'com.gradle.build-scan' version '1.16'
             }
-            
+        """.stripIndent()
+        configureScanPlugin()
+
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        outputContainsExpectedMessage(result.output)
+    }
+
+    def "service messages emitted for compatible Gradle Enterprise plugin versions"() {
+        given:
+        runner.withGradleVersion("6.0-rc-1") // Gradle 6.0 is required for GE plugin 3.0
+        configureBuildScanPlugin(GRADLE_ENTERPRISE_PLUGIN_CLASSPATH)
+        applyPlugin()
+        applyGradleEnterprisePlugin("3.0")
+
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        outputContainsExpectedMessage(result.output)
+    }
+
+    def "no service messages emitted when build scan or Gradle Enterprise plugin is not applied"() {
+        given:
+        configureBuildScanPlugin(BUILD_SCAN_PLUGIN_CLASSPATH)
+        applyPlugin()
+
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        !outputContainsExpectedMessage(result.output)
+    }
+
+    def "no service messages emitted when build scan or Gradle Enterprise plugin are not on the classpath"() {
+        given:
+        configureBuildScanPlugin()
+        applyPlugin()
+
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        !outputContainsExpectedMessage(result.output)
+    }
+
+    private void configureBuildScanPlugin(String classpathSystemPropertyName = null) {
+        runner.withPluginClasspath(PluginUnderTestMetadataReading.readImplementationClasspath() + (classpathSystemPropertyName ? [new File(System.getProperty(classpathSystemPropertyName))] : []))
+    }
+
+    private void configureScanPlugin() {
+        buildScript << """
             buildScan {
                 server = '${mockScansServer.address}'
                 publishAlways()
             }
         """.stripIndent()
     }
 
+    private void applyGradleEnterprisePlugin(String version) {
+        projectDir.newFile("settings.gradle") << """ 
+            plugins {
+                id 'com.gradle.enterprise' version '${version}'
+            }
+            
+            apply plugin: "com.gradle.enterprise"
+            gradleEnterprise {
+                buildScan {
+                    server = '${mockScansServer.address}'
+                    publishAlways()
+                }
+            }
+        """.stripIndent()
+    }
+
+    private void applyPlugin() {
```

Comment:
This should be renamed to something like `applyGradleBuildScanTeamCityPlugin`.

Reply:
done